### PR TITLE
Update shared FFI code

### PR DESCRIPTION
**Pull request by headius**
This PR will track the process of getting files we have in common with the [ffi gem](https://github.com/ffi/ffi) updated so we can share updates and maintenance effort.

See #5947

### CODE REVIEW

File: core/src/main/java/org/jruby/ext/ffi/Platform.java [hunk 1]
```diff
@@ -317,49 +317,12 @@ public static void createPlatformModule(Ruby runtime, RubyModule ffi) {
         OS_TYPE os = platform.getOS();
         module.defineConstant("ADDRESS_SIZE", runtime.newFixnum(platform.addressSize));
         module.defineConstant("LONG_SIZE", runtime.newFixnum(platform.longSize));
-        module.defineConstant("OS", runtime.newString(OS.toString()));
-        module.defineConstant("ARCH", runtime.newString(platform.getCPU().toString()));
-        module.defineConstant("NAME", runtime.newString(platform.getName()));
-        module.defineConstant("IS_WINDOWS", runtime.newBoolean(os == OS.WINDOWS));
-        module.defineConstant("IS_BSD", runtime.newBoolean(platform.isBSD()));
-        module.defineConstant("IS_FREEBSD", runtime.newBoolean(os == OS.FREEBSD));
         module.defineConstant("IS_DRAGONFLYBSD", runtime.newBoolean(os == OS.DRAGONFLYBSD));
```
Comment (ahorek) [inline core/src/main/java/org/jruby/ext/ffi/Platform.java]: https://github.com/ffi/ffi/pull/724

File: lib/ruby/stdlib/ffi/platform/sysconf.rb.ffi [hunk 2]
```diff
@@ -1,139 +0,0 @@
-module Platform;end
```
Comment (ahorek) [inline lib/ruby/stdlib/ffi/platform/sysconf.rb.ffi]: jruby specific?

Comment (headius) [reply lib/ruby/stdlib/ffi/platform/sysconf.rb.ffi]: I think these are template files that originally lived outside of the generation logic but now live in some form in the gen/ dir?

File: core/src/main/java/org/jruby/ext/ffi/Platform.java (hunk 1 above)
Comment (headius) [reply core/src/main/java/org/jruby/ext/ffi/Platform.java]: I nuked all but DragonFly from JRuby's `FFI::Platform` java code. Can nuke DragonFlyBSD once I see it in the ruby code.

**Comment by headius** [comment]
I have had to disable the caching of the Struct layout as part of this PR, but I believe this is due to a misfeature in the C FFI. I have filed an issue here: https://github.com/ffi/ffi/issues/734

**Comment by headius** [comment]
> I have had to disable the caching of the Struct layout

ffi/ffi#734 is fixed now (deprecated behavior to be removed in 2.0) so I reverted the removal of the caching logic.

**Comment by headius** [comment]
Retargeted to 9.3. Help wanted!

File: core/src/main/java/org/jruby/ext/ffi/Platform.java (hunk 1 above)
Comment (larskanis) [reply core/src/main/java/org/jruby/ext/ffi/Platform.java]: @headius `IS_DRAGONFLYBSD` is part of ffi's ruby code now.

**Comment by ahorek** [comment]
the remaining failure is
```
  Struct tests :enum field r/w
     Failure/Error: @orig_field.put(ptr, type.to_native(value, nil))

     NoMethodError:
       undefined method `put' for #<FFI::StructLayout::Number:0x71a509c>
       Did you mean?  puts
                      putc
```

this change https://github.com/jruby/jruby/pull/5948/commits/221b11d3ff9d247d4ca2646be2a68be543d7185a is in conflict with
https://github.com/ffi/ffi/blob/21dc3d48f672ecf334058b36f485292721de8a87/lib/ffi/struct_layout.rb#L92

should we fork the ruby version?

**Comment by headius** [comment]
> should we fork the ruby version?

Ideally, no, we should match the Ruby code exactly.

I'm not sure what you mean by a conflicts. It seems that Field simply doesn't define `put` and `get`. I'm adding those now.

**Comment by ahorek** [comment]
the previous java version used to delegate the put method to MappedFieldIO
https://github.com/jruby/jruby/blob/221b11d3ff9d247d4ca2646be2a68be543d7185a/core/src/main/java/org/jruby/ext/ffi/StructLayout.java#L1417

your fix looks good https://github.com/jruby/jruby/pull/5948/commits/f40efb1270a867278bf3054149ad1d815543db0a thanks

**Comment by headius** [comment]
With these last changes from @ahorek and the missing Field `put` and `get` in place, it appears we are green!

**Comment by headius** [comment]
> the previous java version used to delegate the put method to MappedFieldIO

Ahh yes, but it was never bound in Ruby. All good now!

**Pull request merged**